cmd/mikrotik-codegen/internal/codegen/templates.go

```diff
@@ -0,0 +1,189 @@
+package codegen
+
+const (
+	generatedNotice = "// This code was generated. Review it carefully."
+
+	resourceDefinitionTemplate = `
+package {{ .Package }}
+
+import (
+	{{ range $import := .Imports -}}
+		"{{ $import }}"
+	{{ end }}
+	tftypes "github.com/hashicorp/terraform-plugin-framework/types"
+
+)
+{{ $resourceStructName := .ResourceName | firstLower}}
+type {{$resourceStructName}} struct {
+	client *client.Mikrotik
+}
+
+// Ensure the implementation satisfies the expected interfaces.
+var (
+	_ resource.Resource                = &{{$resourceStructName}}{}
+	_ resource.ResourceWithConfigure   = &{{$resourceStructName}}{}
+	_ resource.ResourceWithImportState = &{{$resourceStructName}}{}
+)
+
+// New{{.ResourceName}}Resource is a helper function to simplify the provider implementation.
+func New{{.ResourceName}}Resource() resource.Resource {
+	return &{{$resourceStructName}}{}
+}
+
+func (r *{{$resourceStructName}}) Configure(_ context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
+
+	r.client = req.ProviderData.(*client.Mikrotik)
+}
+
+// Metadata returns the resource type name.
+func (r *{{$resourceStructName}}) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_{{.ResourceName | lowercase}}"
```

Comment:
I think this should use `ToSnakeCase` instead. I was expecting the `InterfaceWireguard` struct I tested with to result in `mikrotik_interface_wireguard`.

If that's not how you intended it to be used, can you explain how to form the existing resource names?

Reply:
nice catch!
I tested against `single word` resources and didn't even realize it would generate wrong name 😬